core/src/main/java/com/linecorp/armeria/common/util/CachingSupplier.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.util;
+
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.client.limit.ConcurrencyLimit;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * Supplies the cached value that is useful below cases.
+ * <p>
+ * - Supplies dynamically a maximum number of concurrent active requests for {@link ConcurrencyLimit}.
+ */
+public class CachingSupplier<T> implements Supplier<T> {
+    @Nullable
+    private volatile T value;
```

Comment:
How about making this nonnull? because
- We prohibit to return null in the `ConcurrencyLimit`.
- I don't have particular reason/usage to make this nullable. (We can make this nullable later if we find a reason without breaking change)
- Armeria conventionally provides APIs that takes nonnull parameter.

Reply:
Okay I think `nonnull` is more correct.
But I'm not sure if it's good to require that value is not using null `requireNonNull`.

Reply:
I think we should use `requireNonNull` in the `set(...)` and `of(..)` methods If we decide to make this `nonnull` to keep the contract. 😄